### ISSUE TITLE
Blob does not purge

### ISSUE THREAD

**Issue by wbotelhos** [rails/rails#32902]
### Steps to reproduce

1. Create a model with attachment: `has_one_attached :avatar`;
2. Create a validation where if avatar is bigger than you expect it will execute the following code and adds an error on record to avoid the record creation:

```
attachment.purge

errors.add :avatar, 'Too big.'
```

or

```
attachment.purge
attachment.blob.purge

errors.add :avatar, 'Too big.'
```

or

```
attachment.blob.purge
attachment.purge

errors.add :avatar, 'Too big.'
```

or

```
attachment.blob.destroy
attachment.purge

errors.add :avatar, 'Too big.'
```

3. Submit the form with an attached avatar bigger that x to execute the previous code.

### Expected behavior

It should removes the blob from the database.

### Actual behavior

It keeps the blob on database. If the user now makes a valid for with other image, the previous image will be a orphan.

By the way, the sent image comes back and I can render it with `image_tag f.object.avatar`, but if it is sent again, with a valid form, the attachment file is not created and we still have the orphan blob.

### System configuration

**Rails version**: `Rails 5.2.0`

**Ruby version**: `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]`

Thanks for the great work with Rails. (:

**Comment by LikitSaeLee** [comment]
@wbotelhos Hi, does these two test cases describe your problem?
```ruby

  test "when validation failed it should destroy the invalid blob" do
    User.class_eval do
      validate :attachment_size

      def attachment_size
        if avatar.attached? && avatar.blob.byte_size > 100.kilobytes
          avatar.purge
          errors[:avatar] << "Too big"
        end
      end
    end

    valid_avatar = file_fixture "icon.svg"
    oversized_avatar = file_fixture "racecar.jpg"

    @user.update(avatar: Rack::Test::UploadedFile.new(valid_avatar.to_s))

    @user.update(avatar: Rack::Test::UploadedFile.new(oversized_avatar.to_s))

    current_blob_filenames = ActiveStorage::Blob.all.map { |blob| blob.filename.to_s }

    # the blob that didn't pass the validation should be destroyed
    assert_not_includes current_blob_filenames, "racecar.jpg"
  end

  test "when validation failed the original attachment should remain intact" do
    User.class_eval do
      validate :attachment_size

      def attachment_size
        if avatar.attached? && avatar.blob.byte_size > 100.kilobytes
          avatar.purge
          errors[:avatar] << "Too big"
        end
      end
    end

    valid_avatar = file_fixture "icon.svg"
    oversized_avatar = file_fixture "racecar.jpg"

    @user.update(avatar: Rack::Test::UploadedFile.new(valid_avatar.to_s))

    @user.update(avatar: Rack::Test::UploadedFile.new(oversized_avatar.to_s))

    @user.reload

    assert_equal @user.avatar.attached?, true

    assert_equal @user.avatar.blob.filename.to_s, "icon.svg"
  end
```

Strangely, they are both passing for me though, with ruby `2.5.1`. But the `@user.avatar.attachment` will be missing before you do `@user.reload`, is that the problem you encountered?

**Comment by georgeclaghorn** [comment]
Can you provide a minimal sample application demonstrating the issue?

**Comment by CarlosCD** [comment]
Hi @georgeclaghorn, this seems to be related issue #31656, ActiveRecord validations not supported yet. I've seem the same issue, and hopefully I might be able to help.

I am using direct uploads, so when a validation fails I am also left with not attached blobs and files (created via the JS direct upload beforehand), before the attachment happens (no attachment objects created).

As the validations happen within  an ActiveRecord transaction, this means that any intent to delete blobs (purge attachments) would be rollback... a call to purge removes the files (it works), but the blobs remain in the database as a result of the DELETE query rollback.. so the final result would be more inconsistent than doing nothing (I think).

I needed to just postpone the purge to an after_rollback callback. This approach seems to work well... both the files and blobs get removed.

I am using Postgres, Rails 5.2 and direct uploads, and multiple files per object (has_many_attached).

Thanks again for your work!

**Comment by georgeclaghorn** [comment]
Thanks for explaining. This is fixed in master (#33303).
